### PR TITLE
e2e tests

### PR DESCRIPTION
**Pull request by fishkerez**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -25,6 +25,43 @@ jobs:
...
+          docker load --input kubeclarity/kubeclarity.tar
+          docker load --input runtime_k8s_scanner/runtime_k8s_scanner.tar
+          docker load --input kubeclarity-sbom-db/kubeclarity-sbom-db.tar
+          docker load --input cis_docker_benchmark_scanner/cis_docker_benchmark_scanner.tar
+          docker image ls -a
+      - name: Edit values.yaml
+        id: edit_values
+        run: |
+          tag=${{ github.sha }}
+          sed -i 's/latest/'"$tag"'/g' charts/kubeclarity/values.yaml
```
Comment (FrimIdan) [inline .github/workflows/ci.yml:57]: please use the same as in `release.yaml`
```
      find . -type f -name "values.yaml" -o -name "Chart.yaml" |
          xargs sed -i -e s/latest/${{ github.sha }}/g
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -25,6 +25,43 @@ jobs:
...
+          sed -i 's/Always/IfNotPresent/g' charts/kubeclarity/values.yaml
```
Comment (FrimIdan) [inline .github/workflows/ci.yml:58]: why is this needed?

File: .github/workflows/reusable-build-and-push.yml [hunk 3]
```diff
@@ -55,13 +60,20 @@ jobs:
...
+          outputs: type=docker,dest=/tmp/kubeclarity.tar
           cache-from: type=local,src=/tmp/.buildx-cache
           cache-to: type=local,dest=/tmp/.buildx-cache
           build-args: |
             VERSION=${{ inputs.image_tag }}
             BUILD_TIMESTAMP=${{ needs.timestamp.outputs.timestamp }}
             COMMIT_HASH=${{ github.sha }}
 
+      - name: Upload artifact
+        uses: actions/upload-artifact@v3
```
Comment (FrimIdan) [inline .github/workflows/reusable-build-and-push.yml:72]: if: ${{ inputs.upload == 'true' }}

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -25,6 +25,43 @@ jobs:
...
+      - name: Build CLI
+        run: |
+          make cli
+          mv ./cli/bin/cli ./e2e/kubeclarity-cli
+      - name: Run e2e tests
+        run: make e2e
 
   build_all:
     needs: verification
```
Comment (FrimIdan) [inline .github/workflows/ci.yml:67]: need to add to `build_all`
```
with:
  upload: true
```

File: .github/workflows/reusable-build-and-push.yml [hunk 5]
```diff
@@ -55,13 +60,20 @@ jobs:
           tags: ghcr.io/openclarity/kubeclarity:${{ inputs.image_tag }}
           file: Dockerfile.backend
           push: ${{ inputs.push }}
+          outputs: type=docker,dest=/tmp/kubeclarity.tar
```
Comment (FrimIdan) [inline .github/workflows/reusable-build-and-push.yml:63]: need to verify that `push: true` and `outputs` works well together since `push is a is a shorthand for --output=type=registry`

image
from https://github.com/docker/build-push-action


from the code seems like it will fail
https://github.com/docker/buildx/blob/master/commands/build.go#L183

Maybe you can set an env var for output only if `upload` is true and `push` is false and use that env var here. Using and empty env var here might solve it.

example
image

File: charts/kubeclarity/Chart.yaml [hunk 6]
```diff
@@ -2,7 +2,7 @@ apiVersion: v2
 name: kubeclarity
 appVersion: latest
 description: Charts for KubeClarity deployment.
-version: latest
+version: v1.1
```
Comment (FrimIdan) [inline charts/kubeclarity/Chart.yaml:5]: revert

File: e2e/cli_scan_test.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+	"github.com/openclarity/kubeclarity/api/client/client/operations"
+	"github.com/openclarity/kubeclarity/api/client/models"
+	"github.com/openclarity/kubeclarity/e2e/common"
+	"github.com/openclarity/kubeclarity/shared/pkg/formatter"
+)
+
+const (
+	DirectoryAnalyzeOutputSBOMFile = "dir.sbom"
+	ImageAnalyzeOutputSBOMFile     = "merged.sbom"
+	TestImageName                  = "erezfish/test:1.1"
```
Comment (FrimIdan) [inline e2e/cli_scan_test.go:41]: `erezfish` ?

File: e2e/cli_scan_test.go [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+	input := formatter.New(formatter.CycloneDXFormat, sbom)
+	assert.NilError(t, input.Decode(formatter.CycloneDXFormat))
+	return input.GetSBOM().(*cdx.BOM)
+}
+
+func validateAnalyzeDir(t *testing.T) {
+	sbom := getCdxSbom(t, DirectoryAnalyzeOutputSBOMFile)
+	assert.Assert(t, sbom != nil)
+	assert.Assert(t, sbom.Components != nil)
+	//assert.Assert(t, sbom.Metadata.Component.Name == "github.com/cisco-open/kubei/e2e/test_ananlyze")
```
Comment (FrimIdan) [inline e2e/cli_scan_test.go:105]: remove

File: e2e/cli_scan_test.go [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+	assert.Assert(t, len(*sbom.Components) > 0)
+	// TODO assert properties - analyzers = syft, gomod
+}
+
+func validateAnalyzeImage(t *testing.T) {
+	sbom := getCdxSbom(t, ImageAnalyzeOutputSBOMFile)
+	assert.Assert(t, sbom != nil)
+	// check generated sbom
+	assert.Assert(t, sbom.Components != nil)
+	//assert.Assert(t, sbom.Metadata.Component.Name == TestImageName)
```
Comment (FrimIdan) [inline e2e/cli_scan_test.go:115]: remove

File: e2e/common/utils.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+var curDir, _ = os.Getwd()
+var chartPath = filepath.Join(curDir, "../charts/kubeclarity")
+
+func GetCurrentDir() string {
+	return curDir
+}
+
+func InstallKubeClarity(manager *helm.Manager, args string) error {
+	if err := manager.RunInstall(helm.WithName(KubeClarityHelmReleaseName),
+		helm.WithVersion("v1.1"),
```
Comment (FrimIdan) [inline e2e/common/utils.go:81]: `helm.WithVersion("v1.1")` is it a must? if so, the git-sha should be used

File: e2e/common/utils.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+		output, err = cmd.CombinedOutput()
+		if err != nil {
+			processExitedCh <- struct{}{}
+		}
+	}()
+
+	select {
+	case <-stopCh:
+		if err := cmd.Process.Kill(); err != nil {
+			return fmt.Errorf("failed to kill process: %v", err)
```
Comment (FrimIdan) [inline e2e/common/utils.go:188]: do we care? maybe just a print and `return nil` is enough?

File: e2e/runtime_scan_test.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/openclarity/kubeclarity/e2e/common"
+)
+
+func TestRuntimeScan(t *testing.T) {
+	stopCh := make(chan struct{})
+	defer func() {
+		stopCh <- struct{}{}
+		time.Sleep(2 * time.Second)
+	}()
+	f1 := features.New("assert results").
```
Comment (FrimIdan) [inline e2e/runtime_scan_test.go:39]: `assert` ? what is the feature that is being tested?

File: e2e/runtime_scan_test.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+	params := operations.NewPutRuntimeScanStartParams().WithBody(&models.RuntimeScanConfig{
+		Namespaces: namespaces,
+	})
+	_, err := kubeclarityAPI.Operations.PutRuntimeScanStart(params)
+	return err
+}
+
+func waitForScanDone() error {
+	timer := time.NewTimer(3 * time.Minute)
+	ticker := time.NewTicker(3 * time.Second)
```
Comment (FrimIdan) [inline e2e/runtime_scan_test.go:76]: missing defer close

File: e2e/test.yaml [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: test
+    spec:
+      serviceAccountName: test
+      containers:
+        - name: test
+          image: poke/test:latest
```
Comment (FrimIdan) [inline e2e/test.yaml:23]: `poke/test:latest`? maybe we should use `nginx:1.10` or something that is not `latest` and probably will change?

File: e2e/runtime_scan_test.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+func TestRuntimeScan(t *testing.T) {
+	stopCh := make(chan struct{})
+	defer func() {
+		stopCh <- struct{}{}
+		time.Sleep(2 * time.Second)
+	}()
+	f1 := features.New("assert results").
+		WithLabel("type", "assert").
+		Assess("vulnerability in DB", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
```
Comment (FrimIdan) [inline e2e/runtime_scan_test.go:41]: What about testing a failure scan with an image that can't be scanned?

File: e2e/test_analyze/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,17 @@
+FROM golang:1.17.9-alpine AS builder
```
Comment (FrimIdan) [inline e2e/test_analyze/Dockerfile:1]: what is this `test_analyze` folder for?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (fishkerez) [reply .github/workflows/ci.yml:58]: So the docker will load the local image and not try to pull them from remote during e2e tests

File: e2e/common/utils.go (hunk 11 above)
Comment (fishkerez) [reply e2e/common/utils.go:188]: There is no harm in printing this error. It does not fail the test.

File: e2e/test_analyze/Dockerfile (hunk 16 above)
Comment (fishkerez) [reply e2e/test_analyze/Dockerfile]: vulnerable code source and image to scan

File: .github/workflows/ci.yml (hunk 1 above)
Comment (fishkerez) [reply .github/workflows/ci.yml]: Not working for some reason: https://github.com/openclarity/kubeclarity/runs/6558658017?check_suite_focus=true

Comment (FrimIdan) [reply .github/workflows/ci.yml]:
```
find . -type f -name "values.yaml" -o -name "Chart.yaml" | xargs sed -i -e s/latest/41b9c21a81e7fc8618cba764dc7063a9c9329485/g sed -i 's/Always/IfNotPresent/g' charts/kubeclarity/values.yaml
```
`sed -i 's/Always/IfNotPresent/g' charts/kubeclarity/values.yaml` should run in a different line

File: e2e/common/utils.go (hunk 11 above)
Comment (FrimIdan) [reply e2e/common/utils.go:188]: In general it seems odd to me to return an error of a close/kill failure - there is nothing to do with it.

File: e2e/test_analyze/Dockerfile (hunk 16 above)
Comment (FrimIdan) [reply e2e/test_analyze/Dockerfile]: maybe need to add a readme or some comment to understand what is it for

File: e2e/runtime_scan_test.go (hunk 15 above)
Comment (fishkerez) [reply e2e/runtime_scan_test.go]: Done

File: .github/workflows/reusable-build-and-push.yml [hunk 17]
```diff
@@ -48,20 +53,33 @@ jobs:
           username: ${{ github.actor }}
           password: ${{ secrets.GITHUB_TOKEN }}
 
+      - name: Set output env var
```
Comment (akpsgit) [inline .github/workflows/reusable-build-and-push.yml:56]: build output, right?

File: .github/workflows/reusable-build-and-push.yml [hunk 18]
```diff
@@ -126,20 +157,33 @@ jobs:
...
```
Comment (akpsgit) [inline .github/workflows/reusable-build-and-push.yml:160]: same

File: .github/workflows/reusable-build-and-push.yml [hunk 19]
```diff
@@ -165,20 +209,33 @@ jobs:
...
```
Comment (akpsgit) [inline .github/workflows/reusable-build-and-push.yml:212]: same

File: e2e/common/debug.go [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+
+package common
+
+import (
+	"fmt"
+	"os/exec"
+)
+
+func DescribeKubeClarityDeployment() {
+	cmd := exec.Command("kubectl", "-n", "kubeclarity", "describe", "deployments.apps", KubeClarityDeploymentName)
```
Comment (akpsgit) [inline e2e/common/debug.go:24]: KubeClarityNamespace

File: e2e/common/debug.go [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		fmt.Printf("Failed to execute command. %v, %s", err, out)
```
Comment (akpsgit) [inline e2e/common/debug.go:28]: maybe worth printing what command failed

File: e2e/common/debug.go [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+		return
+	}
+	fmt.Printf("kubectl describe deployments.apps -n kubeclarity kubeclarity-kubeclarity:\n %s\n", out)
+}
+
+func DescribeKubeClarityPods() {
+	cmd := exec.Command("kubectl", "-n", "kubeclarity", "describe", "pods")
```
Comment (akpsgit) [inline e2e/common/debug.go:35]: KubeClarityNamespace

File: e2e/common/debug.go [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		fmt.Printf("Failed to execute command. %v, %s", err, out)
+		return
+	}
+	fmt.Printf("kubectl describe pods -n kubeclarity:\n %s\n", out)
+}
+
+func GetKubeClarityPods() {
+	cmd := exec.Command("kubectl", "-n", "kubeclarity", "get", "pods")
```
Comment (akpsgit) [inline e2e/common/debug.go:46]: KubeClarityNamespace

File: e2e/common/utils.go [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+)
+
+// EXPORTED:
+
+func Deploy(ns string, yamlFile string) error {
+	cmd := exec.Command("kubectl", "-n", ns, "apply", "-f", yamlFile)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to execute command. %v, %s", err, out)
```
Comment (akpsgit) [inline e2e/common/utils.go:39]: same

File: e2e/common/utils.go [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+	}
+	return nil
+}
+
+func LoadDockerImagesToCluster(cluster, tag string) error {
```
Comment (akpsgit) [inline e2e/common/utils.go:44]: maybe in each error worth saying what image failed (maybe adding the image names to consts will help)

File: e2e/cli_scan_test.go [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+	t.Helper()
+	dirPath := filepath.Join(common.GetCurrentDir(), "vulnerable")
+
+	command := cliPath + " analyze " + dirPath + " --input-type dir -o " + DirectoryAnalyzeOutputSBOMFile
+
+	cmd := exec.Command("/bin/sh", "-c", command)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Errorf("failed to execute command. %v, %s", err, out)
```
Comment (akpsgit) [inline e2e/cli_scan_test.go:161]: same

File: e2e/cli_scan_test.go [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+	assert.NilError(t, os.Setenv("BACKEND_HOST", "localhost:"+common.KubeClarityPortForwardHostPort))
+	assert.NilError(t, os.Setenv("BACKEND_DISABLE_TLS", "true"))
+
+	command := fmt.Sprintf("%v analyze %v --application-id %v --input-type image --merge-sbom %v -e -o %v", cliPath, TestImageName, appID, inputSbom, ImageAnalyzeOutputSBOMFile)
+
+	cmd := exec.Command("/bin/sh", "-c", command)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Errorf("failed to execute command. %v, %s", err, out)
```
Comment (akpsgit) [inline e2e/cli_scan_test.go:179]: same

File: e2e/cli_scan_test.go [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+	assert.NilError(t, os.Setenv("BACKEND_HOST", "localhost:"+common.KubeClarityPortForwardHostPort))
+	assert.NilError(t, os.Setenv("BACKEND_DISABLE_TLS", "true"))
+
+	command := fmt.Sprintf("%v scan %v --application-id %v --input-type sbom -e", cliPath, inputSbom, appID)
+
+	cmd := exec.Command("/bin/sh", "-c", command)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Errorf("failed to execute command. %v, %s", err, out)
```
Comment (akpsgit) [inline e2e/cli_scan_test.go:194]: same

File: e2e/cli_scan_test.go [hunk 29]
```diff
@@ -0,0 +1,218 @@
...
+	assert.NilError(t, input.Decode(formatter.CycloneDXFormat))
+	return input.GetSBOM().(*cdx.BOM)
+}
+
+func validateAnalyzeDir(t *testing.T) {
+	t.Helper()
+	sbom := getCdxSbom(t, DirectoryAnalyzeOutputSBOMFile)
+	assert.Assert(t, sbom != nil)
+	assert.Assert(t, sbom.Components != nil)
+	//assert.Assert(t, sbom.Metadata.Component.Name == "github.com/cisco-open/kubei/e2e/test_ananlyze")
```
Comment (akpsgit) [inline e2e/cli_scan_test.go:104]: I still see it

File: e2e/runtime_scan_test.go [hunk 30]
```diff
@@ -0,0 +1,121 @@
...
+			t.Logf("get runtime scan results...")
+			results := common.GetRuntimeScanResults(t, kubeclarityAPI)
+			assert.Assert(t, results.Counters.Resources > 0)
+			assert.Assert(t, results.Counters.Vulnerabilities > 0)
+			assert.Assert(t, results.Counters.Packages > 0)
+			assert.Assert(t, results.Counters.Applications > 0)
+
+			assert.Assert(t, len(results.VulnerabilityPerSeverity) > 0)
+
+			assert.Assert(t, len(results.Failures) > 0)
```
Comment (akpsgit) [inline e2e/runtime_scan_test.go:61]: worth adding a comment here. e.g:
radial/busyboxplus:curl is not supported as it uses V1 image manifest, which should result in an error.
